openthread: samples: cli: Enable the sample to substitute OT libs

pszkotak opened the pull request:
The CLI sample can update OpenThread libraries provided by the nrfxlib
with the version build from the current source.

Signed-off-by: Piotr Szkotak <piotr.szkotak@nordicsemi.no>

Note: this PR depends on https://github.com/nrfconnect/sdk-nrfxlib/pull/223.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:38: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#38: FILE: samples/openthread/cli/Kconfig:11:
+        depends on OPENTHREAD_SOURCES$

-:39: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#39: FILE: samples/openthread/cli/Kconfig:12:
+        default n$

-:100: ERROR:TRAILING_WHITESPACE: trailing whitespace
#100: FILE: samples/openthread/cli/common.cmake:24:
+      $

-:102: ERROR:TRAILING_WHITESPACE: trailing whitespace
#102: FILE: samples/openthread/cli/common.cmake:26:
+      $

-:112: ERROR:TRAILING_WHITESPACE: trailing whitespace
#112: FILE: samples/openthread/cli/common.cmake:36:
+  set(OPENTHREAD_CONFIG_FILE $

-:115: ERROR:TRAILING_WHITESPACE: trailing whitespace
#115: FILE: samples/openthread/cli/common.cmake:39:
+  $

-:135: ERROR:TRAILING_WHITESPACE: trailing whitespace
#135: FILE: samples/openthread/cli/common.cmake:59:
+      COMMAND ${CMAKE_COMMAND} -E copy "${lib}" $

-:137: ERROR:TRAILING_WHITESPACE: trailing whitespace
#137: FILE: samples/openthread/cli/common.cmake:61:
+      COMMAND echo $

-:147: ERROR:TRAILING_WHITESPACE: trailing whitespace
#147: FILE: samples/openthread/cli/common.cmake:71:
+    COMMAND ${CMAKE_COMMAND} -E copy $

-:148: ERROR:TRAILING_WHITESPACE: trailing whitespace
#148: FILE: samples/openthread/cli/common.cmake:72:
+      ${NRFXLIB_DIR}/openthread/include/openthread-config-generic.h $

-:150: ERROR:TRAILING_WHITESPACE: trailing whitespace
#150: FILE: samples/openthread/cli/common.cmake:74:
+    COMMAND ${CMAKE_COMMAND} -E remove $

-:152: ERROR:TRAILING_WHITESPACE: trailing whitespace
#152: FILE: samples/openthread/cli/common.cmake:76:
+    COMMAND ${CMAKE_COMMAND} -E remove $

-:154: ERROR:TRAILING_WHITESPACE: trailing whitespace
#154: FILE: samples/openthread/cli/common.cmake:78:
+    COMMAND ${CMAKE_COMMAND} -E remove $

-:156: ERROR:TRAILING_WHITESPACE: trailing whitespace
#156: FILE: samples/openthread/cli/common.cmake:80:
+    COMMAND ${CMAKE_COMMAND} -E remove $

-:158: ERROR:TRAILING_WHITESPACE: trailing whitespace
#158: FILE: samples/openthread/cli/common.cmake:82:
+    COMMAND ${CMAKE_COMMAND} -E remove_directory -f $

-:160: ERROR:TRAILING_WHITESPACE: trailing whitespace
#160: FILE: samples/openthread/cli/common.cmake:84:
+    COMMAND ${CMAKE_COMMAND} -E copy_directory $

-:161: ERROR:TRAILING_WHITESPACE: trailing whitespace
#161: FILE: samples/openthread/cli/common.cmake:85:
+      ${openthread_SOURCE_DIR}/include $

-:163: ERROR:TRAILING_WHITESPACE: trailing whitespace
#163: FILE: samples/openthread/cli/common.cmake:87:
+    COMMAND ${CMAKE_COMMAND} -E copy $

-:164: ERROR:TRAILING_WHITESPACE: trailing whitespace
#164: FILE: samples/openthread/cli/common.cmake:88:
+      ${PROJECT_BINARY_DIR}/openthread-config-generic.h $

- total: 17 errors, 2 warnings, 151 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPDX_LICENSE_TAG SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

rlubos commented on samples/openthread/cli/Kconfig:
```diff
@@ -4,6 +4,17 @@
+config OPENTHREAD_LIBRARIES_SUBSTITUTE
+	bool "Substitute OpenThread Libraries"
+	depends on OPENTHREAD_SOURCES
+	default n
```
No need for `default n`

pszkotak replied:
Fixed.

edmont commented on samples/openthread/cli/Kconfig:
```diff
@@ -4,6 +4,16 @@
+	bool "Substitute OpenThread Libraries"
+	depends on OPENTHREAD_SOURCES
+	help
+	  OpenThread libraries build from source will replace those provided by the nrfxlib.
```
build --> built or compiled

pszkotak replied:
Fixed.

edmont commented on samples/openthread/cli/README.rst:
```diff
@@ -18,6 +18,9 @@ This sample has Commissioner, Joiner, and Border Router roles enabled by default
 It is recommended to use at least two development kits running the same sample to be able to test communication.
 
+An additional functionality of this particular sample is the possibility of updating OpenThread libraries
+with the version build from the current source.
```
build --> built or compiled

pszkotak replied:
Fixed.

edmont commented:
@pszkotak I tried building Thread 1.2 libraries with this method, by applying changes in KRKNWK-6813 and found two issues:

1. Switching to `OPENTHREAD_THREAD_VERSION_1_2` via `west build -t menuconfig` makes `CONFIG_OPENTHREAD_THREAD_VERSION=1.2` but `OT_THREAD_VERSION=1.1`, resulting in Thread 1.1 libraries being built.
2. After setting `CONFIG_OPENTHREAD_THREAD_VERSION_1_2=y` and `CONFIG_OPENTHREAD_LIBRARIES_SUBSTITUTE=y` in the `prj.conf` the Thread 1.1 libraries are not build anymore, but the only result is a single file (not folder): `nrfxlib/openthread/lib/cortex-m4/soft-float/v1.2`.

tejlmand commented on samples/openthread/cli/common.cmake:
```diff
@@ -0,0 +1,97 @@
+      COMMENT "Copying OpenThread library: ${lib}."
+      PRE_LINK
+      COMMAND ${CMAKE_COMMAND} -E copy "${lib}"
+        "${NRFXLIB_DIR}/openthread/${lib_path}/v${OT_THREAD_VERSION}"
```
This is not copying as expected:

Result:
```
$ file openthread/lib/cortex-m4/soft-float/v1.1 
openthread/lib/cortex-m4/soft-float/v1.1: current ar archive
```

pszkotak replied:
I will check that, @edmont reported some copying issues in his tests as well.

pszkotak replied:
Copying works now both for v1.1 and v.1.2. I but we need to hold this PR till architecture is changed to meet behavior described above.

tejlmand commented on samples/openthread/cli/common.cmake:
```diff
@@ -0,0 +1,97 @@
+    COMMENT "Copying OpenThread headers."
+    PRE_LINK
+    COMMAND ${CMAKE_COMMAND} -E copy
+      ${NRFXLIB_DIR}/openthread/include/openthread-config-generic.h
```
This file does not exists.
Have you forgotten a `west.yml` update ?

```
Error copying file "/projects/github/ncs/nrfxlib/openthread/include/openthread-config-generic.h" to "/projects/github/ncs/nrf/samples/openthread/cli/build".
```

Also, why should this file be copied to the `build` folder ?

pszkotak replied:
Yes, `west.yml` is to be updated.
I copy it to preserve it as the whole `${NRFXLIB_DIR}/openthread/include` directory is deleted afterwards.

tejlmand commented on samples/openthread/cli/common.cmake:
```diff
@@ -0,0 +1,97 @@
+# to the nrfxlib openthread folder.
+function(ot_libs_substitute)
+  execute_process(
+    COMMAND           git rev-parse HEAD
```
Use `${GIT_EXECUTABLE}` instead.

Secondly, `git` is not mandatory for building, so please ensure code is working, even if `git` is not available.

pszkotak replied:
Done.

tejlmand commented on samples/openthread/cli/common.cmake:
```diff
@@ -0,0 +1,97 @@
+    COMMAND ${CMAKE_COMMAND} -E copy_directory
+      ${openthread_SOURCE_DIR}/include
+      ${NRFXLIB_DIR}/openthread/include
+    COMMAND ${CMAKE_COMMAND} -E copy
+      ${PROJECT_BINARY_DIR}/openthread-config-generic.h
+      ${NRFXLIB_DIR}/openthread/include
```
I believe you just copied this file in the opposite direction some lines above.

pszkotak replied:
And I deleted the whole directory in the meantime.

tejlmand replied:
so if user had uncommited changes, e.g. new files, then it's just bad luck ?
The build system should not decide to completely remove a folder that could have local changes/new files.

pszkotak replied:
Good point. Implementing a separate target should fix that.

pszkotak commented on samples/openthread/cli/Kconfig:
```diff
@@ -4,6 +4,16 @@
 
+if NET_L2_OPENTHREAD
+
+config OPENTHREAD_LIBRARIES_SUBSTITUTE
+	bool "Substitute OpenThread Libraries"
+	depends on OPENTHREAD_SOURCES
+	help
+	  OpenThread libraries compiled from source will replace those provided by the nrfxlib.
```
> Are end-users expected to replace the libs provided by nrfxlib ?
Yes, there are cases in which users may want to replace libs provided by nrfxlib.

I think the most important one is creating customized version that would be certified and used for other projects.
Provided mechanism makes it very easy.

Another benefit is for us, it makes libraries update process a piece of cake.

>Why do the end-user need to do this while compiling openthread CLI ?
Good question. My initial approach was to provide it as a completely separate example.
I was advised to to combine it with the CLI sample.

It makes sense as the CLI can be used to test pretty much all stack functionalities, so configuration changes made to the OpenThread stack could be implemented and tested with a single build.
Moreover having this functionality implemented as a separate sample seems even more unnatural - it would not even deliver a binary that flashed on the board would do anything useful.

If I understand correctly the SPM does not provide a simple mechanism to replace libraries which is the functionality we need to add here.

tejlmand replied:
> If I understand correctly the SPM does not provide a simple mechanism to replace libraries which is the functionality we need to add here.

Let's keep those two needs separated:
1) Linking application with precompiled libs or build libs from source
2) Copying new libraries to another location.


1 )
---

What i'm adressing, when I refer to SPM is 1).
So the possibility of linking OT cli with pre-compiled libs, or compiled from source should use a similar menu structure.
That is:
```
(X) Link with OT libraries
    (${NRFXLIB_DIR}/openthread/lib/cortex-m4/soft-float/v1.1)      OT library path
( ) Build from source
```
so that we have a uniform way of selecting `Build from source` / `link with lib`.
This furthermore allows end-users to point to a specific lib in case theu want to locate the own version in custom repo.
Just as Nordic tries to avoid patch of upstream Zephyr, then we must expect out users to have similar preference.

`OT library path` is of course the default NRFXLIB path as today.

2 )
---
See comment here: https://github.com/nrfconnect/sdk-nrf/pull/2655#discussion_r469160320

Other
-------

> Good question. My initial approach was to provide it as a completely separate example.
>   I was advised to to combine it with the CLI sample.
>
> It makes sense as the CLI can be used to test pretty much all stack functionalities, so configuration changes made to the  OpenThread stack could be implemented and tested with a single build.

Not sure I follow the rationale, but I'm not advocating for changing this.
I'm just wondering if such approach makes sense to our users.

I see the value in being able to compile from source, but I assume that it would also make sense for light switch to be able to compile from source.
If I understand the workflow correctly, then users would need to do following to use a custom version of OT libs in light switch:

1) Open cli sample
2) build cli with libs from source
3) copy libs
4) open light switch
5) build light switch with new libs

pszkotak replied:
I got the point now. Agree, it makes sense.

1) Actually this has already been implemented: https://github.com/nrfconnect/sdk-nrfxlib/pull/223
The difference between your description and PR mentioned above is the fact that the user is not allowed to enter libs location. If this should be improved we definitely need to work on it in a separate PR.

2) Yes, the workflow you described is exactly the way it works now.
Ticking a menuconfig checkbox `Substitute OpenThread Libraries` is quite explanatory, but having it in the menuconfig may still be misleading.

I will try to implement `ninja install_openthread_libraries` for all OpenThread samples. I agree this is the cleanest solution.

pszkotak commented on samples/openthread/cli/common.cmake:
```diff
@@ -0,0 +1,97 @@
+#
+
+# Header guard
+if(__OT_CLI_COMMON_CMAKE__)
+  return()
+endif()
+set(__OT_CLI_COMMON_CMAKE__ TRUE)
```
You are right.
Is creation a subdirectory with a separate CMakeLists.txt a preffered way to implement a guard here?
As you probably noticed I'm new to CMake, I have no idea how it should be solved *the right way*.

pszkotak replied:
On the other hand I do not see guards in other `.cmake` files.

pszkotak commented on samples/openthread/cli/CMakeLists.txt:
```diff
@@ -6,12 +6,16 @@
 cmake_minimum_required(VERSION 3.13.1)
 
 set(NRF_SUPPORTED_BOARDS
-	nrf52840dk_nrf52840
-	nrf52833dk_nrf52833
```
Removed.

pszkotak commented on samples/openthread/cli/common.cmake:
```diff
@@ -0,0 +1,97 @@
+#
```
Renamed and moved to `nrf/cmake`.

pszkotak commented on samples/openthread/cli/common.cmake:
```diff
@@ -0,0 +1,97 @@
+function(ot_libs_substitute)
+  execute_process(
+    COMMAND           git rev-parse HEAD
+    WORKING_DIRECTORY ${openthread_SOURCE_DIR}
```
What is the best way to ensure it?

tejlmand replied:
From what I can see, compilation of openthread from source is controlled by `CONFIG_OPENTHREAD_SOURCES`:
So this should be a minimum
```
if(CONFIG_OPENTHREAD_SOURCES)
   ...
endif()
```

tejlmand commented on cmake/ot_libs_substitute.cmake:
```diff
@@ -0,0 +1,94 @@
+
+  foreach(lib IN LISTS OPENTHREAD_LIBS)
+    add_custom_command(
+      TARGET app
+      COMMENT "Copying OpenThread library: ${lib}."
+      PRE_LINK
```
We should not copy / overwrite files under version control as part of building the application.

This must be a deliberate action taken by the user.

CMake has a dedicated install target for such operations (which NCS doesn't use, as it make no sense for embedded systems).
But we should follow the principle of having a dedicated target to call when installing the libs.
`ninja install_openthread_libraries`

pszkotak replied:
Agree. Will follow this approach.

edmont reviewed (approved):
Thanks @pszkotak, LGTM :+1: 

This will need to be updated by KRKNWK-6714, but so far it's a great tool.

MarekPorwisz reviewed (commented):
If I am not missing something - ot_libs_configuration_write is called unconditionally. What will happen if we would like to build CLI sample with the use of OT libraries?

pszkotak commented:
> If I am not missing something - ot_libs_configuration_write is called unconditionally. What will happen if we would like to build CLI sample with the use of OT libraries?

It will just create a text file with the configuration and set couple of environmental variables(all `set()` with `CACHE` parameter).
As it is not harmful I decided not to guard it.

edmont commented:
> What will happen if we would like to build CLI sample with the use of OT libraries?

```
CMake Error at /home/edmo/vscode/ncs/nrf/cmake/ot_libs_configuration_write.cmake:14 (execute_process):
  execute_process called with no value for WORKING_DIRECTORY.
Call Stack (most recent call first):
  CMakeLists.txt:17 (ot_libs_configuration_write)
```

pszkotak commented:
> > What will happen if we would like to build CLI sample with the use of OT libraries?
> 
> ```
> CMake Error at /home/edmo/vscode/ncs/nrf/cmake/ot_libs_configuration_write.cmake:14 (execute_process):
>   execute_process called with no value for WORKING_DIRECTORY.
> Call Stack (most recent call first):
>   CMakeLists.txt:17 (ot_libs_configuration_write)
> ```

Fixed and rebased.

rlubos commented:
@tejlmand Could you revisit this PR, the comments are addressed. We have a rare occasion to merge a green PR in case there are no more concerns :)

tejlmand commented on cmake/ot_libs_configuration_write.cmake:
```diff
@@ -0,0 +1,71 @@
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+include(${NRFXLIB_DIR}/common.cmake)
```
This is actually a bad pattern, but I know it's used in nrfxlib also.

But more important. then there is a risk associated here.
This code assumes that `nrfxlib` Zephyr module has been processed first, which may not always be the case.

But you probably noticed that when you made the sample, which is why you have this in the sample.
https://github.com/nrfconnect/sdk-nrf/blob/200decf54c3ef0e5925dde8cd98946134d5f3f68/samples/openthread/cli/CMakeLists.txt#L17

tejlmand commented on samples/openthread/cli/CMakeLists.txt:
```diff
@@ -12,3 +12,28 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+  message(DEBUG "Building OT from sources, config file will be generated.")
+  ot_libs_configuration_write()
+
+  add_custom_target(install_openthread_libraries
```
Much better looking.

The fact that this is an independent target, then why not have it available always, when building Openthread samples ?

Just like you have a `ninja flash`, `ninja run`, `ninja rom_report`, and similar targets available ?

pszkotak replied:
Tried to do it this time. Invoking with `west build -t install_openthread_libraries` works with all openthread examples.

tejlmand replied:
The dependencies are not correctly set up.
If you run:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/openthread/cli/
$ ninja -Cbuild install_openthread_libraries
```
without building the sample, then you will see the following error:
```
FAILED: modules/nrf/CMakeFiles/install_openthread_libraries 
cd /projects/github/ncs/nrf/build/modules/nrf && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/modules/openthread/build/src/core/libopenthread-ftd.a /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/modules/openthread/build/src/core/libopenthread-mtd.a /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/modules/openthread/build/src/core/libopenthread-radio.a /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/modules/openthread/build/src/cli/libopenthread-cli-ftd.a /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/modules/openthread/build/src/cli/libopenthread-cli-mtd.a /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/modules/openthread/build/src/ncp/libopenthread-ncp-ftd.a /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/modules/openthread/build/src/ncp/libopenthread-ncp-mtd.a /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/modules/openthread/build/src/ncp/libopenthread-rcp.a /projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1 && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy_directory /projects/github/ncs/modules/lib/openthread/zephyr/../include /projects/github/ncs/nrfxlib/openthread/include && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E copy /projects/github/ncs/nrf/build/openthread_lib_configuration.txt /projects/github/ncs/nrfxlib/openthread/
Error copying file "/projects/github/ncs/nrf/build/modules/openthread/build/src/core/libopenthread-mtd.a" to "/projects/github/ncs/nrfxlib/openthread/lib/cortex-m4/soft-float/v1.1".
ninja: build stopped: subcommand failed.

```

pszkotak commented on samples/openthread/cli/CMakeLists.txt:
```diff
@@ -12,3 +12,28 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+    COMMAND ${CMAKE_COMMAND} -E make_directory "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_FTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_MTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_RADIO}" "${OPENTHREAD_DST_DIR}"
```
Yes, that's correct. This library will be used by the RCP example which is in our backlog.

tejlmand replied:
thanks for explanation.

pszkotak commented on samples/openthread/cli/CMakeLists.txt:
```diff
@@ -12,3 +12,28 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_CLI_MTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_NCP_FTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_NCP_MTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_RCP}" "${OPENTHREAD_DST_DIR}"
```
Yes, that's correct. This library will be used by the RCP example which is in our backlog.

tejlmand replied:
thanks for explanation.

pszkotak commented on cmake/ot_libs_configuration_write.cmake:
```diff
@@ -0,0 +1,71 @@
+  FILE(WRITE ${OPENTHREAD_CONFIG_FILE} ${OPENTHREAD_SETTINGS})
+
+  set(OPENTHREAD_SRC_DIR "${PROJECT_BINARY_DIR}/modules/openthread/build/src"
+    CACHE STRING "Directory containg the OpenThread source code.")
```
Cached variables are later used by the custom target `install_openthread_libraries`.
It was the only mechanism I could find to make the target executed after the build aware of the paths that it needs to use for copying artifacts.

tejlmand commented on cmake/openthread_extensions.cmake:
```diff
@@ -0,0 +1,121 @@
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+set(NRFXLIB_DIR $ENV{ZEPHYR_BASE}/../nrfxlib)
```
Don't expect `$ENV{ZEPHYR_BASE}`, that is not required to be set.

When referring between module, then `ZEPHYR_<module>_MODULE_DIR` shall be used, in this case, that would be `ZEPHYR_NRFXLIB_MODULE_DIR`.

But in this case, I actually believe this should not be needed in this case, see other comments.

tejlmand commented on cmake/openthread_extensions.cmake:
```diff
@@ -0,0 +1,121 @@
+    CACHE STRING "The nrfxlib directory with OpenThread Libraries to be overwritten.")
+endfunction(openthread_libs_configuration_write)
+
+if(CONFIG_NET_L2_OPENTHREAD AND CONFIG_OPENTHREAD_SOURCES)
+  message(DEBUG "Building OT from sources, config file will be generated.")
+  openthread_libs_configuration_write()
+
+  set(OPENTHREAD_HEADERS_DIR "${ZEPHYR_OPENTHREAD_MODULE_DIR}/../include")
+
+  add_custom_target(install_openthread_libraries
+    COMMAND ${CMAKE_COMMAND} -E make_directory "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_FTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_MTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_RADIO}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_CLI_FTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_CLI_MTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_NCP_FTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_NCP_MTD}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy "${OPENTHREAD_LIB_RCP}" "${OPENTHREAD_DST_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy_directory
+      ${OPENTHREAD_HEADERS_DIR}
+      ${NRFXLIB_DIR}/openthread/include
+    COMMAND ${CMAKE_COMMAND} -E copy
+     ${OPENTHREAD_CONFIG_FILE}
+     ${NRFXLIB_DIR}/openthread/
+    DEPENDS ${OPENTHREAD_LIB_FTD}
+    )
+
+    set_property(TARGET zephyr_property_target
+    APPEND PROPERTY install_openthread_libraries_DEPENDENCIES
+    install_openthread_libraries
+    )
+endif()
```
extension cmake files should only contain extension functions, not code to be executed during CMake configure phase itself.

tejlmand commented on cmake/openthread_extensions.cmake:
```diff
@@ -0,0 +1,121 @@
+    CACHE STRING "File containg OpenThread build configuration parameters.")
+  FILE(WRITE ${OPENTHREAD_CONFIG_FILE} ${OPENTHREAD_SETTINGS})
+
+  set(OPENTHREAD_LIB_FTD "${OPENTHREAD_SRC_DIR}/core/libopenthread-ftd.a"
+    CACHE STRING "FTD")
+  message(STATUS "OPENTHREAD_LIB_FTD=${OPENTHREAD_LIB_FTD}")
+
+  set(OPENTHREAD_LIB_MTD "${OPENTHREAD_SRC_DIR}/core/libopenthread-mtd.a"
+    CACHE STRING "MTD")
+
+  set(OPENTHREAD_LIB_RADIO "${OPENTHREAD_SRC_DIR}/core/libopenthread-radio.a"
+    CACHE STRING "RADIO")
+
+  set(OPENTHREAD_LIB_CLI_FTD "${OPENTHREAD_SRC_DIR}/cli/libopenthread-cli-ftd.a"
+    CACHE STRING "CLI-FTD")
+
+  set(OPENTHREAD_LIB_CLI_MTD "${OPENTHREAD_SRC_DIR}/cli/libopenthread-cli-mtd.a"
+    CACHE STRING "CLI-MTD")
+
+  set(OPENTHREAD_LIB_NCP_FTD "${OPENTHREAD_SRC_DIR}/ncp/libopenthread-ncp-ftd.a"
+    CACHE STRING "NCP-FTD")
+
+  set(OPENTHREAD_LIB_NCP_MTD "${OPENTHREAD_SRC_DIR}/ncp/libopenthread-ncp-mtd.a"
+    CACHE STRING "NCP-MTD")
+
+  set(OPENTHREAD_LIB_RCP "${OPENTHREAD_SRC_DIR}/ncp/libopenthread-rcp.a"
+    CACHE STRING "RCP")
```
Please remove:
```suggestion
```

tejlmand commented on cmake/openthread_extensions.cmake:
```diff
@@ -0,0 +1,121 @@
+  unset(OPENTHREAD_LIB_RCP CACHE)
+  unset(OPENTHREAD_DST_DIR CACHE)
+
+  message(STATUS "OPENTHREAD_SRC_DIR=${OPENTHREAD_SRC_DIR}")
```
Please remove, the CMake configure phase very limited in what it should print of information.
Info, such as: Toolchain, Zephyr base, board info, and similar.

tejlmand commented on cmake/openthread_extensions.cmake:
```diff
@@ -0,0 +1,121 @@
+  message(STATUS "OPENTHREAD_SRC_DIR=${OPENTHREAD_SRC_DIR}")
+  set(OPENTHREAD_SRC_DIR "${CMAKE_BINARY_DIR}/modules/openthread/build/src"
+    CACHE STRING "Directory containg the OpenThread source code.")
+  message(STATUS "CMAKE_BINARY_DIR=${CMAKE_BINARY_DIR}")
+  message(STATUS "OPENTHREAD_SRC_DIR=${OPENTHREAD_SRC_DIR}")
```
Please remove.
```suggestion
```

tejlmand commented on cmake/openthread_extensions.cmake:
```diff
@@ -0,0 +1,121 @@
+
+  find_package(Git QUIET)
+  if(GIT_FOUND)
+    execute_process(
+      COMMAND           ${GIT_EXECUTABLE} rev-parse HEAD
+      WORKING_DIRECTORY ${OPENTHREAD_SRC_DIR}
```
This will only work if the openthread module CMake code has been executed BEFORE this code.
Otherwise, `OPENTHREAD_SRC_DIR` may not be present.

In my test, I get:
```
cat build/openthread_lib_configuration.txt 
OpenThread_commit=
```

Also, you are getting the commits SHA of the build folder, I don't believe that is your intention.
The build folder can by anywere, even in `/tmp`

tejlmand commented on cmake/openthread_extensions.cmake:
```diff
@@ -0,0 +1,121 @@
+# Store the configuration of the compiled OpenThread libraries
+# and set source and destination paths.
+function(openthread_libs_configuration_write)
+  unset(OPENTHREAD_SRC_DIR CACHE)
+  unset(OPENTHREAD_CONFIG_FILE CACHE)
+  unset(OPENTHREAD_LIB_FTD CACHE)
+  unset(OPENTHREAD_LIB_MTD CACHE)
+  unset(OPENTHREAD_LIB_RADIO CACHE)
+  unset(OPENTHREAD_LIB_CLI_FTD CACHE)
+  unset(OPENTHREAD_LIB_CLI_MTD CACHE)
+  unset(OPENTHREAD_LIB_NCP_FTD CACHE)
+  unset(OPENTHREAD_LIB_NCP_MTD CACHE)
+  unset(OPENTHREAD_LIB_RCP CACHE)
+  unset(OPENTHREAD_DST_DIR CACHE)
```
please remove
```suggestion
```

tejlmand reviewed (changes requested):
Still some work to do, but direction is getting better:
See some suggestions in this commit, that you can squash in:
https://github.com/tejlmand/fw-nrfconnect-nrf/commit/c2259e0faaf66a48c15f046798ddbb2dad125de7

tejlmand commented on cmake/openthread_extensions.cmake:
```diff
@@ -0,0 +1,121 @@
+    DEPENDS ${OPENTHREAD_LIB_FTD}
+    )
+
+    set_property(TARGET zephyr_property_target
+    APPEND PROPERTY install_openthread_libraries_DEPENDENCIES
+    install_openthread_libraries
```
Care to explain why this is needed ?

pszkotak replied:
You are right, it was not needed. Already removed it.

tejlmand commented on samples/openthread/cli/README.rst:
```diff
@@ -61,6 +64,8 @@ Building and running
 
 .. include:: /includes/build_and_run.txt
 
+To update OpenThread libraries provided by the ``nrfxlib`` with the currently built ones please invoke ``ninja -C build/ install_openthread_libraries`` after building the CLI example from source.
```
If dependencies are specified correctly, then it is enough to just invoke:
```
ninja -C build/ install_openthread_libraries
```

No need to do a CLI build first.
If libraries are not built, then running the target will ensure they gets built and installed.
If they are already built, they will just be installed.

```suggestion
To update OpenThread libraries provided by the ``nrfxlib``, please invoke ``ninja -C build/ install_openthread_libraries``.
```

pszkotak replied:
As discussed added modified version with the `west` instead of the `ninja` command.

tejlmand commented:
@pszkotak  Could you trigger a rebuild, to ensure we get CI to run including the latest nrfxlib update ?

pszkotak commented:
> @pszkotak Could you trigger a rebuild, to ensure we get CI to run including the latest nrfxlib PR update ?

Sure, done.

tejlmand commented:
please update west.yml to SHA: 7ab7c9f1e4cc2d55bb71ec9116dfcad7efb540f2

pszkotak commented:
> please update west.yml to SHA: 7ab7c9f1e4cc2d55bb71ec9116dfcad7efb540f2

Done. Thanks for your guidance in this PR.

lmaciejonczyk commented:
@rlubos All seams done. Can we merge it?

greg-fer commented:
Guys, I guess you missed a tech writer for this PR... not good, not good...

tejlmand commented:
and merged when PR still has a `DNM` label :stuck_out_tongue:
